src/include/compat.h

```diff
@@ -307,4 +307,12 @@ int lchown(const char *path, uid_t owner, gid_t group);
 
 #endif /* WIN32 */
 
+#ifdef __MINGW32__
+#define CEPH_PACKED(DECL) DECL __attribute__((packed, gcc_struct))
```

Comment:
I'll give it a try, thanks a lot for the links.

We're not using msys2 at the moment. Ubuntu 18.04 was the latest Ubuntu LTS version when we wrote the patches and it provides mingw gcc 7.3. Since other people that are building/testing this code are also using older versions, we stuck to that.

Reply:
Ok, so that's a gcc 4.7 regression that was fixed in gcc 8. Basically the ``packed`` attribute was broken by ``-mms-bitfields``, even for structures that don't have any bitfields.

We have a few options:
* use ``-mno-ms-bitfields`` - while avoiding the ``packed`` issue, we may lose ABI compatibility with Windows DLLs as well as other 3rd party DLLs, which I'd rather avoid
* keep this macro but properly document it, saying that it's just for backwards compatibility with older Mingw versions, especially meant for structures that are passed between hosts.
* drop ``CEPH_PACKED`` and require recent Mingw versions. We might document this and even add a build script check so that people won't have to spend time debugging it. Compile time assertions would also bring it up.

Probably the last option would be the preferred one so that we don't clutter the Ceph codebase.

Reply:
I dropped this, documenting Mingw 8 as the minimum required version.